provider/oci/images.go

```diff
@@ -0,0 +1,403 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package oci
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/juju/errors"
+	// jujuos "github.com/juju/utils/os"
+	"github.com/juju/utils/series"
+
+	"github.com/juju/juju/environs/imagemetadata"
+	"github.com/juju/juju/environs/instances"
+	"github.com/juju/juju/provider/oci/common"
+
+	ociCore "github.com/oracle/oci-go-sdk/core"
+)
+
+const (
+	BareMetal      InstanceType = "metal"
+	VirtualMachine InstanceType = "vm"
+	GPUMachine     InstanceType = "gpu"
+
+	// ImageTypeVM should be run on a virtual instance
+	ImageTypeVM ImageType = "vm"
+	// ImageTypeBM should be run on bare metal
+	ImageTypeBM ImageType = "metal"
+	// ImageTypeGPU should be run on an instance with attached GPUs
+	ImageTypeGPU ImageType = "gpu"
+	// ImageTypeGeneric should work on any type of instance (bare metal or virtual)
+	ImageTypeGeneric ImageType = "generic"
+
+	windowsOS = "Windows"
+	centOS    = "CentOS"
+	ubuntuOS  = "Canonical Ubuntu"
+
+	staleImageCacheTimeoutInMinutes = 30
+)
+
+var globalImageCache = &ImageCache{}
+var cacheMutex = &sync.Mutex{}
+
+type InstanceType string
+type ImageType string
+
+type ImageVersion struct {
+	TimeStamp time.Time
+	Revision  int
+}
+
+func setImageCache(cache *ImageCache) {
+	cacheMutex.Lock()
+	defer cacheMutex.Unlock()
+
+	globalImageCache = cache
+}
+
+func NewImageVersion(img ociCore.Image) (ImageVersion, error) {
+	var imgVersion ImageVersion
+	if img.DisplayName == nil {
+		return imgVersion, errors.Errorf("image does not have a display name")
+	}
+	fields := strings.Split(*img.DisplayName, "-")
+	if len(fields) < 2 {
+		return imgVersion, errors.Errorf("invalid image display name")
```

Comment:
Can we include the actual img.Display name in this return?
Seeing it in tests makes it clearer that if a user ever saw this error, they would have a really hard time understanding *what* was actually wrong.
errors.Errorf("invalid image display name %q", *img.DisplayName)
should be reasonable.